### ISSUE TITLE
Question: CloudFormation size limit

### ISSUE THREAD

**Issue by jennycld** [awslabs/aws-service-catalog-puppet#634]
We have an Organization Unit that has 60+ accounts, we need to share some portfolios/products to these accounts using the factory/puppet pipeline. We listed all 60+ accounts in the manifest.yaml file and got an error saying "'templateBody' failed to satisfy constraint: Member must have length less than or equal to 51200". If we reduce the number of accounts in the manifest.yaml file to about 30, we are able to share the portfolios/products to the puppet accounts. 

My understanding is the framework reads the manifest.yaml file and generates a CFN template, the more accounts we have, the bigger the CFN template, eventually the CFN template reaches the size limitation of 51,200 bytes and the pipeline fails. 

I understand instead of listing all 60+ accounts, we can use OU id or path in the manifest.yaml file, this certainly reduce the size of the manifest.yaml file. The question is, does this also reduce the size of the CFN template? Will we still run into the same size limitation issue since the framework will get a list of all accounts in the OU and expands the manifest.yaml file? If we run into the same size limitation issue, what is your recommendation on how to resolve it? 

Thank you

**Comment by eamonnfaherty** [comment]
Hey @jennycld thanks for raising an issue.

There are a few AWS CloudFormation templates that vary in size depending on the number of accounts you have defined.  

Could you please tell me which AWS CloudFormation stack fails to create/update with the 60+ accounts.  

Also, sharing your manifest file with me would help resolve the issue more quickly.

**Comment by jennycld** [comment]
Hi @eamonnfaherty, 

Thanks for your prompt response. 

The pipeline seems to have failed for CloudFormation stack servicecatalog-puppet-policies, in build project servicecatalog-puppet-deploy, when running command "servicecatalog-puppet --info deploy --num-workers ${NUM_WORKERS} manifest-expanded.yaml". It seems for each account, the pipeline adds 4 blocks of CFN code (for TopicPolicies, BucketPolicies, BucketPoliciesForRepo and EventBusPolicy) to the CloudFormation template, so the more accounts we have, the bigger the CFN template gets, until it reaches the size limit. 

Attached please see a copy of the manifest file I used, with account IDs and account names replaced with <acct_id> and <acct_name>. In the manifest file there are 34 accounts, if we add one more account the pipeline will fail. We are considering using OU path/name in the manifest file instead of listing all individual accounts, but if the pipeline expands the OU and still lists all individual accounts in the manifest-expaned.yaml file, I am afraid we would run into the same size limitation issue. 

Please let me know if you need any other info. 

Thank you.
Jenny

[manifest.yaml.zip](https://github.com/awslabs/aws-service-catalog-puppet/files/10724628/manifest.yaml.zip)

**Comment by jennycld** [comment]
Hi @eamonnfaherty,

I noticed another behavior and would like to share with you. 

We have multiple OUs in our Organization, during my last testing, I listed 2 OUs in the manifest.yaml file, these two OUs only have a small number of accounts so I don't have to worry about exceeding the size limit for CFN. After the pipeline finished successfully, I checked the servicecatalog-puppet-policies, it seems instead of creating the 4 resources (TopicPolicies, BucketPolicies, BucketPoliciesForRepo and EventBusPolicy) for the 2 listed OUs, it created the 4 resources for the whole Organization. This will be a problem because we have certain portfolios that we would only like to share with 2 listed OUs, not the rest of the OUs in the Org. 

snippet from the servicecatalog-puppet-policies template: 

  TopicPolicies:
    Type: AWS::SNS::TopicPolicy
    Properties:
      <Omit Topics: …>
      PolicyDocument:
        Id: MyTopicPolicy
        Version: '2012-10-17'
        Statement: 
        
          - Sid: **OrganizationalShareForo-MyOrgID <- share for the whole Org**
            Action:
              - sns:Publish
            Effect: "Allow"
            Resource: "*"
            Principal: "*"
            Condition:
              StringEquals:
                **aws:PrincipalOrgID: o-MyOrgID <- share for the whole Org, shouldn’t this only be the 2 listed OUs?**  

Thanks, 
Jenny
